Cython/Compiler/ExprNodes.py

```diff
@@ -12221,7 +12221,11 @@ def result_type(self, type1, type2, env):
                 type1 = Builtin.bytes_type
             elif type1.is_pyunicode_ptr:
                 type1 = Builtin.unicode_type
-            if type1.is_builtin_type or type2.is_builtin_type:
+            if type1 is Builtin.type_type or type2 is Builtin.type_type:
+                # This only really applies to | - the result is a subclass of 'type'
+                # but not an exact instance of 'type'.
+                return py_object_type 
```

Comment:
Rather than adding a special special-case here in the base class, I'd remove `|` from the "safe operators" below and implement `BitwiseOrNode.infer_builtin_types_operation()` appropriately.